File: Balance_BST.cpp
class Solution {
public:
    vector<int>res;
    void inorder(TreeNode* root){
        if(!root){return;}
        inorder(root->left);
        res.push_back(root->val);
        inorder(root->right);
    }
    TreeNode* makeBST(int start,int end){
        if(start>end){
            return NULL;
        }
        int mid = (start+end)/2;
        TreeNode*root = new TreeNode(res[mid]);

        root->left = makeBST(start,mid-1);
        root->right = makeBST(mid+1,end);

        return root;

    }
    TreeNode* balanceBST(TreeNode* root) {
        inorder(root);
        int start =0;
        int end = res.size()-1;
        return makeBST(start,end);
    }
};

File: Balance_BST.py
class Solution:
    def __init__(self):
        self.res = []

    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        self.res.append(root.val)
        self.inorder(root.right)

    def makeBST(self, start, end):
        if start > end:
            return None
        mid = (start + end) // 2
        root = TreeNode(self.res[mid])
        root.left = self.makeBST(start, mid - 1)
        root.right = self.makeBST(mid + 1, end)
        return root

    def balanceBST(self, root):
        self.res = []  # Ensure res is reset for each new call
        self.inorder(root)
        start = 0
        end = len(self.res) - 1
        return self.makeBST(start, end)
